kafka-ui-api/src/main/java/com/provectus/kafka/ui/model/FailoverUrlList.java

```diff
@@ -6,55 +6,53 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
 
 public class FailoverUrlList {
 
   public static final int DEFAULT_RETRY_GRACE_PERIOD_IN_MS = 5000;
 
-  private final Map<Integer, Instant> failures = new HashMap<>();
+  private final Map<Integer, Instant> failures = new ConcurrentHashMap<>();
+  private final AtomicInteger index = new AtomicInteger(0);
   private final List<String> urls;
   private final int retryGracePeriodInMs;
 
-  private Integer index;
-
   public FailoverUrlList(List<String> urls) {
     this(urls, DEFAULT_RETRY_GRACE_PERIOD_IN_MS);
   }
 
   public FailoverUrlList(List<String> urls, int retryGracePeriodInMs) {
     if (urls != null && !urls.isEmpty()) {
       this.urls = new ArrayList<>(urls);
-      this.index = new Random().nextInt(urls.size());
     } else {
       throw new IllegalArgumentException("Expected at least one URL to be passed in constructor");
     }
     this.retryGracePeriodInMs = retryGracePeriodInMs;
   }
 
   public String current() {
-    return this.urls.get(this.index);
+    return this.urls.get(this.index.get());
   }
 
   public synchronized void fail(String url) {
```

Comment:
Now, you dont need to use synchronized here. ConcurrentHashMap guarantees visibility, and only mutation operation is compareAndSet which is thread-safe